src/quick-lint-js/parse.h

```diff
@@ -2515,8 +2515,26 @@ class parser {
     }
 
     if (this->peek().type == token_type::kw_else) {
+      source_code_span else_span = this->peek().span();
       this->skip();
-      parse_and_visit_body();
+
+      switch (this->peek().type) {
+      default:
+        parse_and_visit_body();
+        break;
+
+      case token_type::left_paren:
+        expression *ast = this->parse_expression(precedence{});
+        this->visit_expression(ast, v, variable_context::rhs);
+
+        if (this->peek().type == token_type::left_curly) {
```

Comment:
I think we also need to check the kind of `ast`. See examples in my 'blocking' comment.

In #266, I wrote "side-effect-free expression". I think this is conservative but probably still helpful. "side-effect-free" would be any combination of:
* `binary_operator` (I think?) (if children are side-effect free)
* `cond`
* `_typeof`
* `_class`
* `template` (if children are side-effect free)
* `array` (if children are side-effect free)
* `object` (if children are side-effect free)
* `conditional` (if children are side-effect free)
* ...

In particular:
* `dot` is disallowed because property access can have arbitrary side effects

A `bool is_side_effect_free(expression*)` function would be a welcome addition!

Reply:
I'm not sure yet about all the base cases, see body of `bool is_side_effect_free(expression* ast)`.

Reply:
As for checking the kind of the expression, right now the only invalidating kinds are arrow functions. I'm trying to come up with other examples where it's not caught by the side-effect-free condition.